pkg/state/commitments.go

```diff
@@ -0,0 +1,150 @@
+package state
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/fxamacker/cbor/v2"
+
+	"github.com/wavesplatform/gowaves/pkg/crypto"
+	"github.com/wavesplatform/gowaves/pkg/crypto/bls"
+	"github.com/wavesplatform/gowaves/pkg/keyvalue"
+	"github.com/wavesplatform/gowaves/pkg/proto"
+)
+
+// commitmentItem represents a single commitment made by a block generator.
+// It links the generator's Waves public key to its corresponding BLS endorser public key.
+type commitmentItem struct {
+	GeneratorPK crypto.PublicKey `cbor:"0,keyasint,omitempty"`
+	EndorserPK  bls.PublicKey    `cbor:"1,keyasint,omitempty"`
+}
+
+// commitmentsRecord holds all generator commitments for a specific generation period.
+type commitmentsRecord struct {
+	Commitments []commitmentItem `cbor:"0,keyasint,omitempty"`
+}
+
+func (cr *commitmentsRecord) append(generatorPK crypto.PublicKey, endorserPK bls.PublicKey) {
+	cr.Commitments = append(cr.Commitments, commitmentItem{
+		GeneratorPK: generatorPK,
+		EndorserPK:  endorserPK,
+	})
+}
+func (cr *commitmentsRecord) marshalBinary() ([]byte, error) { return cbor.Marshal(cr) }
+
+func (cr *commitmentsRecord) unmarshalBinary(data []byte) error { return cbor.Unmarshal(data, cr) }
+
+// commitments manages the storage and retrieval of generator commitments.
+type commitments struct {
+	hs *historyStorage
+}
+
+func newCommitments(hs *historyStorage) *commitments {
+	return &commitments{
+		hs: hs,
+	}
+}
+
+func (c *commitments) store(
+	periodStart uint32, generatorPK crypto.PublicKey, endorserPK bls.PublicKey, blockID proto.BlockID,
+) error {
+	key := commitmentKey{periodStart: periodStart}
+	data, err := c.hs.newestTopEntryData(key.bytes())
+	if err != nil && !errors.Is(err, keyvalue.ErrNotFound) {
```

Comment:
If you're looking for data in history use `isNotFoundInHistoryOrDBErr` helper function.

Reply:
Fixed